src/Clients/FSpot/FSpot/GtkSynchronizationContext.cs

```diff
@@ -0,0 +1,147 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Pulled from here: https://github.com/mono/monodevelop/blob/master/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/DispatchService.cs#L48
+
+using System;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+using Hyena;
+
+namespace FSpot
+{
+    class GtkSynchronizationContext : SynchronizationContext
+    {
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        internal delegate bool GSourceFuncInternal(IntPtr ptr);
+
+        internal class TimeoutProxy
+        {
+            internal SendOrPostCallback d;
+            object state;
+            ManualResetEventSlim resetEvent;
+
+            public TimeoutProxy(SendOrPostCallback d, object state) : this(d, state, null)
+            {
+            }
+
+            public TimeoutProxy(SendOrPostCallback d, object state, ManualResetEventSlim lockObject)
+            {
+                this.d = d;
+                this.state = state;
+                this.resetEvent = lockObject;
+            }
+
+            internal static readonly GSourceFuncInternal SourceHandler = HandlerInternal;
+
+            static bool HandlerInternal(IntPtr data)
+            {
+                var proxy = (TimeoutProxy)((GCHandle)data).Target;
+
+                try
+                {
+                    proxy.d(proxy.state);
+                }
+                catch (Exception e)
+                {
+                    GLib.ExceptionManager.RaiseUnhandledException(e, false);
+                }
+                finally
+                {
+                    proxy.resetEvent?.Set();
+                }
+                return false;
+            }
+        }
+
+        const int defaultPriority = 0;
+
+        [DllImport("libglib-2.0-0.dll", CallingConvention = CallingConvention.Cdecl)]
+        static extern uint g_timeout_add_full(int priority, uint interval, GSourceFuncInternal d, IntPtr data, GLib.DestroyNotify notify);
+
+        class ExceptionWithStackTraceWithoutThrowing : Exception
+        {
+            readonly string stacktrace;
+
+            public ExceptionWithStackTraceWithoutThrowing(string message) : base(message)
+            {
+                stacktrace = Environment.StackTrace;
+            }
+
+            public override string StackTrace => stacktrace;
+        }
+
+        static void AddTimeout(TimeoutProxy proxy)
+        {
+            if (proxy.d == null)
```

Comment:
You don't need this, this was added to check a bug in MD. Should be removed there too.